assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js

```diff
@@ -0,0 +1,343 @@
+	useMemo,
+} from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
+import { CORE_LOCATION } from '../../../../googlesitekit/datastore/location/constants';
+import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
+import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
+import { EDIT_SCOPE as ANALYTICS_EDIT_SCOPE } from '../../../analytics/datastore/constants';
+import {
+	FORM_CUSTOM_DIMENSIONS_CREATE,
+	MODULES_ANALYTICS_4,
+} from '../../datastore/constants';
+import { KEY_METRICS_WIDGETS } from '../../../../components/KeyMetrics/key-metrics-widgets';
+import { Button } from 'googlesitekit-components';
+import Link from '../../../../components/Link';
+import MetricTileError from '../../../../components/KeyMetrics/MetricTileError';
+import MetricTileLoading from '../../../../components/KeyMetrics/MetricTileLoading';
+import {
+	ERROR_CODE_MISSING_REQUIRED_SCOPE,
+	ERROR_REASON_BAD_REQUEST,
+	isInsufficientPermissionsError,
+} from '../../../../util/errors';
+import { useFeature } from '../../../../hooks/useFeature';
+import { getWidgetComponentProps } from '../../../../googlesitekit/widgets/util';
+const { useSelect, useDispatch } = Data;
+
+export default function CustomDimensionsWrapper( {
+	children,
+	reportError,
+	widgetSlug,
+	title = KEY_METRICS_WIDGETS[ widgetSlug ]?.title,
+	infoTooltip = KEY_METRICS_WIDGETS[ widgetSlug ]?.infoTooltip ||
+		KEY_METRICS_WIDGETS[ widgetSlug ]?.description,
+} ) {
+	const newsKeyMetricsEnabled = useFeature( 'newsKeyMetrics' );
+
+	const customDimensions = useMemo( () => {
+		if ( ! widgetSlug || ! newsKeyMetricsEnabled ) {
+			return null;
+		}
+
+		const { requiredCustomDimensions } = KEY_METRICS_WIDGETS[ widgetSlug ];
+
+		if (
+			! Array.isArray( requiredCustomDimensions ) ||
+			! requiredCustomDimensions.length
+		) {
+			return null;
+		}
+
+		return requiredCustomDimensions;
+	}, [ newsKeyMetricsEnabled, widgetSlug ] );
+
+	const hasCustomDimensions = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return true;
+		}
+
+		return select( MODULES_ANALYTICS_4 ).hasCustomDimensions(
+			customDimensions
+		);
+	} );
+	const isCreatingCustomDimensions = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return false;
+		}
+
+		return customDimensions.some( ( dimension ) =>
+			select( MODULES_ANALYTICS_4 ).isCreatingCustomDimension( dimension )
+		);
+	} );
+	const customDimensionsCreationErrors = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return [];
+		}
+
+		return customDimensions
+			.filter(
+				( dimension ) =>
+					!! select(
+						MODULES_ANALYTICS_4
+					).getCreateCustomDimensionError( dimension )
+			)
+			.map( ( dimension ) =>
+				select( MODULES_ANALYTICS_4 ).getCreateCustomDimensionError(
+					dimension
+				)
+			);
```

Comment:
We call the same selector twice per each item. This is inefficient. Let's convert this into classical for loop.

Reply:
Updated, does this look better now? Thanks!